File: ExclusiveTimeFunction.java
//Time Complexity: O(n)
//Space Complexity: O(n)
public class ExclusiveTimeFunction {
    public int[] exclusiveTime(int n, List<String> logs)
    {
        if(logs == null || logs.size() == 0)
            return new int[]{};

        int result[] = new int[n];
        Stack<Integer> st = new Stack<>();
        int prev = 0;
        for(String log : logs)
        {
            // get all vals from string
            String [] arr = log.split(":");
            
            int curr = Integer.parseInt(arr[2]);

            if(arr[1].equals("start"))
            {
                // compute how long earlier process ran
                if(!st.isEmpty()) // first task will not have earlier process
                {
                    result[st.peek()] = curr - prev;
                }
                // update old timestamp
                prev = curr;
                // new function
                int fid = Integer.parseInt(arr[0]);
                st.push(fid);
            }
            else if(arr[1].equals("end")) 
            {
                // update run time for earlier process and since end we can pop
                int fid = st.pop();
                result[fid] += curr + 1 - prev; // need 1 for current timestamp
                prev = curr + 1; // add single unit timestamp as 1
            }
        }
        return result;
    }
}

File: exclusiveTimeFunction.py
'''
Time Complexity: O(n)
Space Complexity:O(n)
Did this code successfully run on Leetcode : Yes
Explanation: Create a stack and if we see a start for the function - >push the functionId into the stack while saving
the currentTimestamp - priorTimestamp into the result array and if we see an end we pop the top of stack (as it contains
the currently executing function) and update the time its been running by currentTimestamp - priorTimestamp + 1 and
update the priorTimestamp by currentTimestamp+1
'''
class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        stack = []
        res = [0 for i in range(n)]
        priorTimestamp = 0
        for log in logs:
            token = log.split(':')
            functionId = int(token[0])
            currentTimestamp = int(token[2])
            if token[1] == "start":
                if len(stack) != 0:
                    res[stack[-1]] += currentTimestamp - priorTimestamp
                stack.append(functionId)
                priorTimestamp = currentTimestamp
            else:
                res[stack.pop()] += currentTimestamp - priorTimestamp + 1
                priorTimestamp = currentTimestamp + 1

        return res